File: sdks/typescript/src/v1/inbound.ts
import type {
  AttachmentMetaView,
  AttachmentView,
  Authentication,
  MessageView,
  SendResultView,
} from "./generated/index.js";
import { E2AValidationError } from "./errors.js";
import type { ForwardInput, ReplyInput, RequestOptions } from "./client.js";
import type { EmailReceivedData, WebhookEvent } from "./webhook-signature.js";

export type EmailReceivedEvent = WebhookEvent & {
  type: "email.received";
  schema_version: "1";
  data: EmailReceivedData;
};

export interface InboundMessageOperations {
  get(email: string, id: string): Promise<MessageView>;
  getAttachment(
    email: string,
    id: string,
    index: number,
    options?: { inline?: boolean },
  ): Promise<AttachmentView>;
  reply(
    email: string,
    id: string,
    body: ReplyInput,
    options?: RequestOptions,
  ): Promise<SendResultView>;
  forward(
    email: string,
    id: string,
    body: ForwardInput,
    options?: RequestOptions,
  ): Promise<SendResultView>;
}

export interface InboundAttachmentProjection {
  content_id?: string;
  content_type?: string;
  filename?: string;
  index: number;
  size_bytes: number;
}

export interface InboundProjection {
  id: string;
  inbox: string;
  conversation_id: string;
  from: string | null;
  envelope_from: string | null;
  verified: boolean;
  to: string[];
  cc: string[];
  reply_to: string[];
  reply_targets: string[];
  subject: string;
  text: string;
  html?: string;
  text_truncated: boolean;
  received_at: string;
  flagged: boolean;
  flag_reason?: string;
  attachments: InboundAttachmentProjection[];
}

function invalidEvent(message: string): E2AValidationError {
  return new E2AValidationError({
    code: "invalid_email_received_event",
    message,
    status: 0,
    retryable: false,
  });
}

function isNonEmptyString(value: unknown): value is string {
  return typeof value === "string" && value.length > 0;
}

function validateEvent(event: WebhookEvent): EmailReceivedEvent {
  if (!event || typeof event !== "object") {
    throw invalidEvent("expected an email.received event object");
  }
  if (event.schema_version !== "1" || event.type !== "email.received") {
    throw invalidEvent("expected a schema-v1 email.received event");
  }
  if (!event.data || typeof event.data !== "object" || Array.isArray(event.data)) {
    throw invalidEvent("email.received data must be an object");
  }
  const data = event.data as Record<string, unknown>;
  if (!isNonEmptyString(data.message_id) || !isNonEmptyString(data.delivered_to)) {
    throw invalidEvent("email.received requires non-empty message_id and delivered_to fetch keys");
  }
  return event as EmailReceivedEvent;
}

function attachmentProjection(meta: AttachmentMetaView): InboundAttachmentProjection {
  return {
    ...(meta.contentId === undefined ? {} : { content_id: meta.contentId }),
    ...(meta.contentType === undefined ? {} : { content_type: meta.contentType }),
    ...(meta.filename === undefined ? {} : { filename: meta.filename }),
    index: meta.index,
    size_bytes: meta.sizeBytes,
  };
}

export class InboundAttachment {
  readonly contentId?: string;
  readonly contentType?: string;
  readonly filename?: string;
  readonly index: number;
  readonly sizeBytes: number;

  constructor(
    meta: AttachmentMetaView,
    private readonly inbox: string,
    private readonly messageId: string,
    private readonly messages: InboundMessageOperations,
  ) {
    this.contentId = meta.contentId;
    this.contentType = meta.contentType;
    this.filename = meta.filename;
    this.index = meta.index;
    this.sizeBytes = meta.sizeBytes;
  }

  get(options: { inline?: boolean } = {}): Promise<AttachmentView> {
    return this.messages.getAttachment(this.inbox, this.messageId, this.index, options);
  }

  toJSON(): InboundAttachmentProjection {
    return attachmentProjection(this);
  }
}

export class InboundEmail {
  readonly id: string;
  readonly inbox: string;
  readonly conversationId: string;
  readonly from: string | null;
  readonly envelopeFrom: string | null;
  readonly authentication: Authentication | null;
  readonly verified: boolean;
  readonly to: readonly string[];
  readonly cc: readonly string[];
  readonly replyTo: readonly string[];
  readonly replyTargets: readonly string[];
  readonly subject: string;
  readonly text: string;
  readonly html?: string;
  readonly textTruncated: boolean;
  readonly receivedAt: Date;
  readonly flagged: boolean;
  readonly flagReason?: string;
  readonly attachments: readonly InboundAttachment[];

  constructor(
    readonly event: EmailReceivedEvent,
    readonly message: MessageView,
    private readonly messages: InboundMessageOperations,
  ) {
    this.id = message.id;
    this.inbox = message.deliveredTo;
    this.conversationId = message.conversationId;
    this.from = message.headerFrom;
    this.envelopeFrom = message.envelopeFrom;
    this.authentication = message.authentication;
    this.verified = message.authentication?.dmarc.status === "pass";
    this.to = Object.freeze([...message.to]);
    this.cc = Object.freeze([...message.cc]);
    this.replyTo = Object.freeze([...message.replyTo]);
    this.replyTargets = Object.freeze(
      message.replyTo.length > 0 ? [...message.replyTo] : message.headerFrom === null ? [] : [message.headerFrom],
    );
    this.subject = message.subject;
    this.text = message.parsed?.text ?? "";
    this.html = message.parsed?.html;
    this.textTruncated = message.parsed?.truncated ?? false;
    this.receivedAt = new Date(message.createdAt);
    this.flagged = message.flagged ?? false;
    this.flagReason = message.flagReason;
    this.attachments = Object.freeze(message.attachments.map(
      (meta) => new InboundAttachment(meta, this.inbox, this.id, messages),
    ));
  }

  reply(body: ReplyInput, options: RequestOptions = {}): Promise<SendResultView> {
    return this.messages.reply(this.inbox, this.id, body, options);
  }

  forward(body: ForwardInput, options: RequestOptions = {}): Promise<SendResultView> {
    return this.messages.forward(this.inbox, this.id, body, options);
  }

  toJSON(): InboundProjection {
    return {
      id: this.id,
      inbox: this.inbox,
      conversation_id: this.conversationId,
      from: this.from,
      envelope_from: this.envelopeFrom,
      verified: this.verified,
      to: [...this.to],
      cc: [...this.cc],
      reply_to: [...this.replyTo],
      reply_targets: [...this.replyTargets],
      subject: this.subject,
      text: this.text,
      ...(this.html === undefined ? {} : { html: this.html }),
      text_truncated: this.textTruncated,
      received_at: this.receivedAt.toISOString(),
      flagged: this.flagged,
      ...(this.flagReason === undefined ? {} : { flag_reason: this.flagReason }),
      attachments: this.attachments.map((attachment) => attachment.toJSON()),
    };
  }
}

export class InboundResource {
  constructor(private readonly messages: InboundMessageOperations) {}

  async fromEvent(event: WebhookEvent): Promise<InboundEmail> {
    const received = validateEvent(event);
    const message = await this.messages.get(received.data.delivered_to, received.data.message_id);
    return new InboundEmail(received, message, this.messages);
  }
}

File: sdks/python/src/e2a/v1/inbound.py
"""High-level inbound email facade over verified event envelopes."""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Any, Coroutine, List, Mapping, Optional, Protocol, TypeVar

from .errors import E2AValidationError
from .generated.models import (
    AttachmentMetaView,
    AttachmentView,
    Authentication,
    MessageView,
    SendResultView,
)

__all__ = [
    "InboundEvent",
    "AsyncInboundResource",
    "AsyncInboundEmail",
    "AsyncInboundAttachment",
    "InboundResource",
    "InboundEmail",
    "InboundAttachment",
]


class InboundEvent(Protocol):
    """Structural webhook/WebSocket envelope accepted by the facade."""

    @property
    def id(self) -> str: ...

    @property
    def type(self) -> str: ...

    @property
    def schema_version(self) -> str: ...

    @property
    def created_at(self) -> str: ...

    @property
    def data(self) -> Any: ...


class InboundMessageOperations(Protocol):
    async def get(self, email: str, message_id: str) -> MessageView: ...

    async def get_attachment(
        self, email: str, message_id: str, index: int, *, inline: bool = False
    ) -> AttachmentView: ...

    async def reply(
        self,
        email: str,
        message_id: str,
        body: Any,
        *,
        idempotency_key: Optional[str] = None,
    ) -> SendResultView: ...

    async def forward(
        self,
        email: str,
        message_id: str,
        body: Any,
        *,
        idempotency_key: Optional[str] = None,
    ) -> SendResultView: ...


def _invalid_event(message: str) -> E2AValidationError:
    return E2AValidationError(
        code="invalid_email_received_event",
        message=message,
        status=0,
        retryable=False,
    )


def _validated_data(event: InboundEvent) -> Mapping[str, Any]:
    if event.schema_version != "1" or event.type != "email.received":
        raise _invalid_event("expected a schema-v1 email.received event")
    if not isinstance(event.data, Mapping):
        raise _invalid_event("email.received data must be a mapping")
    message_id = event.data.get("message_id")
    delivered_to = event.data.get("delivered_to")
    if not isinstance(message_id, str) or not message_id:
        raise _invalid_event("email.received requires a non-empty message_id fetch key")
    if not isinstance(delivered_to, str) or not delivered_to:
        raise _invalid_event("email.received requires a non-empty delivered_to fetch key")
    return event.data


def _iso_milliseconds(value: datetime) -> str:
    normalized = value
    if normalized.tzinfo is None:
        normalized = normalized.replace(tzinfo=timezone.utc)
    normalized = normalized.astimezone(timezone.utc)
    return normalized.isoformat(timespec="milliseconds").replace("+00:00", "Z")


def _attachment_dict(meta: AttachmentMetaView) -> dict[str, Any]:
    result: dict[str, Any] = {
        "index": meta.index,
        "size_bytes": meta.size_bytes,
    }
    if meta.content_id is not None:
        result["content_id"] = meta.content_id
    if meta.content_type is not None:
        result["content_type"] = meta.content_type
    if meta.filename is not None:
        result["filename"] = meta.filename
    return result


class AsyncInboundAttachment:
    def __init__(
        self,
        meta: AttachmentMetaView,
        inbox: str,
        message_id: str,
        messages: InboundMessageOperations,
    ) -> None:
        self.content_id = meta.content_id
        self.content_type = meta.content_type
        self.filename = meta.filename
        self.index = meta.index
        self.size_bytes = meta.size_bytes
        self._inbox = inbox
        self._message_id = message_id
        self._messages = messages

    async def get(self, *, inline: bool = False) -> AttachmentView:
        return await self._messages.get_attachment(
            self._inbox, self._message_id, self.index, inline=inline
        )

    def to_dict(self) -> dict[str, Any]:
        return _attachment_dict(self)  # type: ignore[arg-type]

    def __repr__(self) -> str:
        return f"AsyncInboundAttachment({self.to_dict()!r})"


class AsyncInboundEmail:
    def __init__(
        self,
        event: InboundEvent,
        message: MessageView,
        messages: InboundMessageOperations,
    ) -> None:
        data = _validated_data(event)
        self.event = event
        self.message = message
        self.id = message.id
        self.inbox = message.delivered_to
        self.conversation_id = message.conversation_id
        self.from_ = message.header_from
        self.envelope_from = message.envelope_from
        self.authentication: Optional[Authentication] = message.authentication
        self.verified = bool(
            message.authentication and message.authentication.dmarc.status == "pass"
        )
        self.to = list(message.to)
        self.cc = list(message.cc)
        self.reply_to = list(message.reply_to)
        self.reply_targets = (
            list(message.reply_to)
            if message.reply_to
            else ([message.header_from] if message.header_from is not None else [])
        )
        self.subject = message.subject
        self.text = message.parsed.text if message.parsed is not None else ""
        self.html = message.parsed.html if message.parsed is not None else None
        self.text_truncated = message.parsed.truncated if message.parsed is not None else False
        self.received_at = message.created_at
        self.flagged = bool(message.flagged)
        self.flag_reason = message.flag_reason
        self.attachments = [
            AsyncInboundAttachment(meta, self.inbox, self.id, messages)
            for meta in message.attachments
        ]
        self._messages = messages

    async def reply(
        self, body: Any, *, idempotency_key: Optional[str] = None
    ) -> SendResultView:
        return await self._messages.reply(
            self.inbox, self.id, body, idempotency_key=idempotency_key
        )

    async def forward(
        self, body: Any, *, idempotency_key: Optional[str] = None
    ) -> SendResultView:
        return await self._messages.forward(
            self.inbox, self.id, body, idempotency_key=idempotency_key
        )

    def to_dict(self) -> dict[str, Any]:
        result: dict[str, Any] = {
            "id": self.id,
            "inbox": self.inbox,
            "conversation_id": self.conversation_id,
            "from": self.from_,
            "envelope_from": self.envelope_from,
            "verified": self.verified,
            "to": list(self.to),
            "cc": list(self.cc),
            "reply_to": list(self.reply_to),
            "reply_targets": list(self.reply_targets),
            "subject": self.subject,
            "text": self.text,
            "text_truncated": self.text_truncated,
            "received_at": _iso_milliseconds(self.received_at),
            "flagged": self.flagged,
            "attachments": [attachment.to_dict() for attachment in self.attachments],
        }
        if self.html is not None:
            result["html"] = self.html
        if self.flag_reason is not None:
            result["flag_reason"] = self.flag_reason
        return result

    def __repr__(self) -> str:
        return f"AsyncInboundEmail({self.to_dict()!r})"


class AsyncInboundResource:
    def __init__(self, messages: InboundMessageOperations) -> None:
        self._messages = messages

    async def from_event(self, event: InboundEvent) -> AsyncInboundEmail:
        data = _validated_data(event)
        message = await self._messages.get(data["delivered_to"], data["message_id"])
        return AsyncInboundEmail(event, message, self._messages)


T = TypeVar("T")


class _Bridge(Protocol):
    def submit(self, coro: Coroutine[Any, Any, T]) -> T: ...


class InboundAttachment:
    """Blocking adapter over :class:`AsyncInboundAttachment`."""

    def __init__(self, target: AsyncInboundAttachment, bridge: _Bridge) -> None:
        self._target = target
        self._bridge = bridge
        self.content_id = target.content_id
        self.content_type = target.content_type
        self.filename = target.filename
        self.index = target.index
        self.size_bytes = target.size_bytes

    def get(self, *, inline: bool = False) -> AttachmentView:
        return self._bridge.submit(self._target.get(inline=inline))

    def to_dict(self) -> dict[str, Any]:
        return self._target.to_dict()

    def __repr__(self) -> str:
        return f"InboundAttachment({self.to_dict()!r})"


class InboundEmail:
    """Blocking adapter over :class:`AsyncInboundEmail`."""

    def __init__(self, target: AsyncInboundEmail, bridge: _Bridge) -> None:
        self._target = target
        self._bridge = bridge
        self.event = target.event
        self.message = target.message
        self.id = target.id
        self.inbox = target.inbox
        self.conversation_id = target.conversation_id
        self.from_ = target.from_
        self.envelope_from = target.envelope_from
        self.authentication = target.authentication
        self.verified = target.verified
        self.to = list(target.to)
        self.cc = list(target.cc)
        self.reply_to = list(target.reply_to)
        self.reply_targets = list(target.reply_targets)
        self.subject = target.subject
        self.text = target.text
        self.html = target.html
        self.text_truncated = target.text_truncated
        self.received_at = target.received_at
        self.flagged = target.flagged
        self.flag_reason = target.flag_reason
        self.attachments = [InboundAttachment(item, bridge) for item in target.attachments]

    def reply(
        self, body: Any, *, idempotency_key: Optional[str] = None
    ) -> SendResultView:
        return self._bridge.submit(
            self._target.reply(body, idempotency_key=idempotency_key)
        )

    def forward(
        self, body: Any, *, idempotency_key: Optional[str] = None
    ) -> SendResultView:
        return self._bridge.submit(
            self._target.forward(body, idempotency_key=idempotency_key)
        )

    def to_dict(self) -> dict[str, Any]:
        return self._target.to_dict()

    def __repr__(self) -> str:
        return f"InboundEmail({self.to_dict()!r})"


class InboundResource:
    """Blocking adapter over :class:`AsyncInboundResource`."""

    def __init__(self, target: AsyncInboundResource, bridge: _Bridge) -> None:
        self._target = target
        self._bridge = bridge

    def from_event(self, event: InboundEvent) -> InboundEmail:
        target = self._bridge.submit(self._target.from_event(event))
        return InboundEmail(target, self._bridge)
